python/src/nanoarrow/visitor.py

```diff
@@ -15,68 +15,187 @@
-        :func:`c_array_stream`.
-    schema : schema-like, optional
-        An optional schema, passed to :func:`c_array_stream`.
+def nulls_forbid() -> Callable[[CBuffer, Sequence], Sequence]:
+    """Erroring null handler
+
+    A null handler that errors when it encounters nulls.
 
     Examples
     --------
 
     >>> import nanoarrow as na
-    >>> from nanoarrow import visitor
-    >>> array = na.c_array([1, 2, 3], na.int32())
-    >>> visitor.to_pylist(array)
-    [1, 2, 3]
+    >>> na.Array([1, 2, 3], na.int32()).to_column(na.nulls_forbid())
+    nanoarrow.c_lib.CBuffer(int32[12 b] 1 2 3)
+    >>> na.Array([1, None, 3], na.int32()).to_column(na.nulls_forbid())
+    Traceback (most recent call last):
+    ...
+    ValueError: Null present with null_handler=nulls_forbid()
     """
-    return ListBuilder.visit(obj, schema)
 
+    def handle(is_valid, data):
+        if len(is_valid) > 0:
+            raise ValueError("Null present with null_handler=nulls_forbid()")
+
+        return data
 
-def to_columns(obj, schema=None) -> Tuple[List[str], List[Sequence]]:
-    """Convert ``obj`` to a ``list()` of sequences
+    return handle
 
-    Converts a stream of struct arrays into its column-wise representation
-    such that each column is either a contiguous buffer or a ``list()``.
 
-    Paramters
-    ---------
-    obj : array stream-like
-        An array-like or array stream-like object as sanitized by
-        :func:`c_array_stream`.
-    schema : schema-like, optional
-        An optional schema, passed to :func:`c_array_stream`.
+def nulls_as_sentinel(sentinel=None):
+    """Sentinel null handler
+
+    A null handler that assigns a sentinel to null values. This is
+    done using numpy using the expression ``data[~is_valid] = sentinel``.
+    The default sentinel value will result in ``nan`` assigned to null
+    values in numeric and boolean outputs.
+
+    Parameters
+    ----------
+    sentinel : scalar, optional
+        The value with which nulls should be replaced.
 
     Examples
     --------
 
     >>> import nanoarrow as na
+    >>> na.Array([1, 2, 3], na.int32()).to_column(na.nulls_as_sentinel())
+    array([1, 2, 3], dtype=int32)
+    >>> na.Array([1, None, 3], na.int32()).to_column(na.nulls_as_sentinel())
+    array([ 1., nan,  3.])
+    >>> na.Array([1, None, 3], na.int32()).to_column(na.nulls_as_sentinel(-999))
+    array([   1, -999,    3], dtype=int32)
+    """
+    import numpy as np
+
+    def handle(is_valid, data):
+        is_valid = np.array(is_valid, copy=False)
+        data = np.array(data, copy=False)
+
+        if len(is_valid) > 0:
+            out_type = np.result_type(data, sentinel)
+            data = np.array(data, dtype=out_type, copy=True)
+            data[~is_valid] = sentinel
+            return data
+        else:
+            return data
+
+    return handle
+
+
+def nulls_debug() -> Callable[[CBuffer, Sequence], Tuple[CBuffer, Sequence]]:
```

Comment:
Should we mark this as an internal API? e.g. `_nulls_debug()`

Reply:
`nulls_separate` sounds in any case as a better name! (and yes I think it would be useful)